src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/TdsParserHelperClasses.cs

```diff
@@ -511,73 +511,130 @@ public object Clone()
-                for (int idx = 0; idx < metaDataArray.Length; idx++)
+                _metaDataArray = new _SqlMetaData[original._metaDataArray.Length];
+                for (int idx = 0; idx < _metaDataArray.Length; idx++)
                 {
-                    metaDataArray[idx] = (_SqlMetaData)original.metaDataArray[idx].Clone();
+                    _metaDataArray[idx] = (_SqlMetaData)original._metaDataArray[idx].Clone();
                 }
             }
         }
 
+        internal int VisibleColumnCount
+        {
+            get
+            {
+                if (_hiddenColumnCount == -1)
+                {
+                    SetupHiddenColumns();
+                }
+                return Length - _hiddenColumnCount;
+            }
+        }
+
         internal int Length
         {
             get
             {
-                return metaDataArray.Length;
+                return _metaDataArray.Length;
             }
         }
 
         internal _SqlMetaData this[int index]
         {
             get
             {
-                return metaDataArray[index];
+                return _metaDataArray[index];
             }
             set
             {
                 Debug.Assert(null == value, "used only by SqlBulkCopy");
-                metaDataArray[index] = value;
+                _metaDataArray[index] = value;
             }
         }
 
-        public object Clone()
+        public int GetVisibleColumnIndex(int index)
+        {
+            if (_hiddenColumnCount == -1)
+            {
+                SetupHiddenColumns();
+            }
+            if (_visibleColumnMap is null)
+            {
+                return index;
+            }
+            else
+            {
+                return _visibleColumnMap[index];
+            }
+        }
+
+        public _SqlMetaDataSet Clone()
         {
             return new _SqlMetaDataSet(this);
         }
+
+        private void SetupHiddenColumns()
+        {
+            int hiddenColumnCount = 0;
+            for (int index = 0; index < Length; index++)
+            {
+                if (_metaDataArray[index].IsHidden)
+                {
+                    hiddenColumnCount += 1;
+                }
+            }
+
+            if (hiddenColumnCount > 0)
+            {
+                int[] visibleColumnMap = new int[Length - hiddenColumnCount];
+                int mapIndex = 0;
+                for (int metaDataIndex = 0; metaDataIndex < Length; metaDataIndex++)
+                {
+                    if (!_metaDataArray[metaDataIndex].IsHidden)
+                    {
+                        visibleColumnMap[mapIndex] = metaDataIndex;
+                        mapIndex += 1;
+                    }
+                }
+                _visibleColumnMap = visibleColumnMap;
+            }
+            _hiddenColumnCount = hiddenColumnCount;
+        }
     }
 
+
     sealed internal class _SqlMetaDataSetCollection : ICloneable
```

Comment:
A minor change from the PR [#521](https://github.com/dotnet/SqlClient/pull/521/files#diff-a001e9005c2d64819a2b2139c27a807a2c45c18bbaa9db2d51f0f5f1a301435eR710), which could be ported.